File: main/shared/instructions/instructions.cc
/*
 * MIT License
 *
 * Copyright (c) 2023 Cunyuan(Holden) Gao
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include "instructions.h"

namespace instructions {
  bool InstABC::Is(uint p_code) {
    switch (p_code & OP_CODE_MASK) {
      case static_cast<int>(OpCode::MOVE):
      case static_cast<int>(OpCode::LOAD_FALSE):
      case static_cast<int>(OpCode::L_FALSE_SKIP):
      case static_cast<int>(OpCode::LOAD_TRUE):
      case static_cast<int>(OpCode::LOAD_NIL):
      case static_cast<int>(OpCode::GET_UP_VAL):
      case static_cast<int>(OpCode::SET_UP_VAL):
      case static_cast<int>(OpCode::GET_TAB_UP):
      case static_cast<int>(OpCode::GET_TABLE):
      case static_cast<int>(OpCode::GET_I):
      case static_cast<int>(OpCode::GET_FIELD):
      case static_cast<int>(OpCode::SET_TAB_UP):
      case static_cast<int>(OpCode::SET_TABLE):
      case static_cast<int>(OpCode::SET_I):
      case static_cast<int>(OpCode::SET_FIELD):
      case static_cast<int>(OpCode::NEW_TABLE):
      case static_cast<int>(OpCode::SELF):
      case static_cast<int>(OpCode::ADD_I):
      case static_cast<int>(OpCode::ADD_K):
      case static_cast<int>(OpCode::SUB_K):
      case static_cast<int>(OpCode::MUL_K):
      case static_cast<int>(OpCode::MOD_K):
      case static_cast<int>(OpCode::POW_K):
      case static_cast<int>(OpCode::DIV_K):
      case static_cast<int>(OpCode::I_DIV_K):
      case static_cast<int>(OpCode::B_AND_K):
      case static_cast<int>(OpCode::B_OR_K):
      case static_cast<int>(OpCode::B_XOR_K):
      case static_cast<int>(OpCode::SHR_I):
      case static_cast<int>(OpCode::SHL_I):
      case static_cast<int>(OpCode::ADD):
      case static_cast<int>(OpCode::SUB):
      case static_cast<int>(OpCode::MUL):
      case static_cast<int>(OpCode::MOD):
      case static_cast<int>(OpCode::POW):
      case static_cast<int>(OpCode::DIV):
      case static_cast<int>(OpCode::I_DIV):
      case static_cast<int>(OpCode::B_AND):
      case static_cast<int>(OpCode::B_OR):
      case static_cast<int>(OpCode::B_XOR):
      case static_cast<int>(OpCode::SHL):
      case static_cast<int>(OpCode::SHR):
      case static_cast<int>(OpCode::MM_BIN):
      case static_cast<int>(OpCode::MM_BIN_I):
      case static_cast<int>(OpCode::MM_BIN_K):
      case static_cast<int>(OpCode::UNM):
      case static_cast<int>(OpCode::B_NOT):
      case static_cast<int>(OpCode::NOT):
      case static_cast<int>(OpCode::LEN):
      case static_cast<int>(OpCode::CONCAT):
      case static_cast<int>(OpCode::CLOSE):
      case static_cast<int>(OpCode::TBC):
      case static_cast<int>(OpCode::EQ):
      case static_cast<int>(OpCode::LT):
      case static_cast<int>(OpCode::LE):
      case static_cast<int>(OpCode::EQ_K):
      case static_cast<int>(OpCode::EQ_I):
      case static_cast<int>(OpCode::LT_I):
      case static_cast<int>(OpCode::LE_I):
      case static_cast<int>(OpCode::GT_I):
      case static_cast<int>(OpCode::GE_I):
      case static_cast<int>(OpCode::TEST):
      case static_cast<int>(OpCode::TEST_SET):
      case static_cast<int>(OpCode::CALL):
      case static_cast<int>(OpCode::TAIL_CALL):
      case static_cast<int>(OpCode::RETURN):
      case static_cast<int>(OpCode::RETURN_0):
      case static_cast<int>(OpCode::RETURN_1):
      case static_cast<int>(OpCode::T_FOR_CALL):
      case static_cast<int>(OpCode::SET_LIST):
      case static_cast<int>(OpCode::VAR_ARG):
      case static_cast<int>(OpCode::VAR_ARG_PREP):
        return true;
      default:
        return false;
    }
  }

  InstABC::InstABC(uint p_code):
    code(static_cast<OpCode>(p_code & OP_CODE_MASK)),
    a(static_cast<uchar>((p_code >> 7) & 0xFF)), k(((p_code >> 15) & 1) > 0),
    b(static_cast<uchar>((p_code >> 16) & 0xFF)),
    c(static_cast<uchar>((p_code >> 24) & 0xFF)) {}

  bool InstABx::Is(uint p_code) {
    switch (p_code & OP_CODE_MASK) {
      case static_cast<int>(OpCode::LOAD_K):
      case static_cast<int>(OpCode::LOAD_KX):
      case static_cast<int>(OpCode::FOR_LOOP):
      case static_cast<int>(OpCode::FOR_PREP):
      case static_cast<int>(OpCode::T_FOR_PREP):
      case static_cast<int>(OpCode::T_FOR_LOOP):
      case static_cast<int>(OpCode::CLOSURE):
        return true;
      default:
        return false;
    }
  }

  InstABx::InstABx(uint p_code):
    code(static_cast<OpCode>(p_code & OP_CODE_MASK)),
    a(static_cast<uchar>((p_code >> 7) & 0xFF)),
    bx(static_cast<ushort>((p_code >> 15) & 0x1FFFF)) {}

  bool InstAsBx::Is(uint p_code) {
    switch (p_code & OP_CODE_MASK) {
      case static_cast<int>(OpCode::LOAD_I):
      case static_cast<int>(OpCode::LOAD_F):
        return true;
      default:
        return false;
    }
  }

  InstAsBx::InstAsBx(uint p_code):
    code(static_cast<OpCode>(p_code & OP_CODE_MASK)),
    a(static_cast<uchar>((p_code >> 7) & 0xFF)),
    sbx(static_cast<short>((p_code >> 15) & 0x1FFFF)) {}

  bool InstAx::Is(uint p_code) {
    return (p_code & OP_CODE_MASK) == static_cast<int>(OpCode::EXTRA_ARG);
  }

  InstAx::InstAx(uint p_code):
    code(static_cast<OpCode>(p_code & OP_CODE_MASK)),
    ax(static_cast<uint>((p_code >> 15) & 0x1FFFFFF)) {}

  bool InstsJ::Is(uint p_code) {
    return (p_code & OP_CODE_MASK) == static_cast<int>(OpCode::JMP);
  }

  InstsJ::InstsJ(uint p_code):
    code(static_cast<OpCode>(p_code & OP_CODE_MASK)),
    sj(static_cast<uint>((p_code >> 15) & 0x1FFFFFF)) {}

  std::variant<Instruction, Error> Parse(uint p_code) {
    if (InstABC::Is(p_code)) {
      return InstABC{p_code};
    } else if (InstABx::Is(p_code)) {
      return InstABx{p_code};
    } else if (InstAsBx::Is(p_code)) {
      return InstAsBx{p_code};
    } else if (InstAx::Is(p_code)) {
      return InstAx{p_code};
    } else if (InstsJ::Is(p_code)) {
      return InstsJ{p_code};
    } else {
      return "wrong operation code.";
    }
  }
} // namespace instructions

File: generator/instructions.ml
(*
 * MIT License
 *
 * Copyright (c) 2023 Cunyuan(Holden) Gao
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *)

open Cpplib

(* put instructions in order *)
(* TODO: add more check for C++ AST *)
let instructions = [
  InstABC ( (* 0 *)
    (fun ins -> ["Move"]),
    (fun ins -> [
      const "value" (get_stack (a ins));
      const "pos" (b ins);
      stmt (set_stack "pos" "value");
      return("1")
    ])
  );
  InstAsBx (
    (fun ins -> ["Load I at %d, value: %d"; a(ins); sbx(ins)]),
    (fun ins -> [
      stmt (set_stack (a ins) (sbx ins));
      return("1")
    ])
  );
  InstAsBx (
    (fun ins -> ["Load F at %d, value: %f"; a(ins); "static_cast<double>(" ^ sbx(ins) ^ ")"]),
    (fun ins -> [
      stmt (set_stack (a ins) (static_cast "double" (sbx ins)));
      return("1")
    ])
  );
  InstABx (
    (fun ins -> ["Load K(constant) from %d to %d"; bx(ins); a(ins)]),
    (fun ins -> [
      stmt (push (from_lit (get_const (bx ins))));
      stmt (replace_top (a ins));
      return("1")
    ])
  );
  InstABx (
    (fun ins -> ["Load KX"]),
    (fun ins -> [return("1")])
  );
  InstABC ( (* 5 *)
    (fun ins -> ["Load False at %d"; a(ins)]),
    (fun ins -> [
      stmt (set_stack (a ins) "false");
      return("1")
    ])
  );
  InstABC (
    (fun ins -> ["L False Skip"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Load True at %d"; a(ins)]),
    (fun ins -> [
      stmt (set_stack (a ins) "true");
      return("1")
    ])
  );
  InstABC (
    (fun ins -> ["Load Nil: [%d, %d]"; a(ins); a(ins) ^ " + " ^ b(ins)]),
    (fun ins -> [
      stmt (push null);
      for_0_le_inc (b ins) [stmt(copy_stack "-1" "i")];
      stmt (pop);
      return("1")
    ])
  );
  InstABC (
    (fun ins -> ["Get Up Value"]),
    (fun ins -> [return("1")])
  );
  InstABC ( (* 10 *)
    (fun ins -> ["Set Up Value"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Get Table Up"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Get Table"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Get I"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Get Field"]),
    (fun ins -> [return("1")])
  );
  InstABC ( (* 15 *)
    (fun ins -> ["Set Table Up"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Set Table"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Set I"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Set Field"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["New Table"]),
    (fun ins -> [return("1")])
  );
  InstABC ( (* 20 *)
    (fun ins -> ["Self"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Add I"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Add K"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Sub K"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Mul K"]),
    (fun ins -> [return("1")])
  );
  InstABC ( (* 25 *)
    (fun ins -> ["Mod K"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Pow K"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Div K"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["I Div K"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Bit And K"]),
    (fun ins -> [return("1")])
  );
  InstABC ( (* 30 *)
    (fun ins -> ["Bit Or K"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Bit Xor K"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Shift Right I"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Shift Left I"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Add R[%d] = R[%d] + R[%d]"; a(ins); b(ins); c(ins)]),
    (fun ins -> [
      "const auto p1 = p_stack->Get(" ^ b(ins) ^ ");";
      "const auto p2 = p_stack->Get(" ^ c(ins) ^ ");";
      "const auto res = CalcArith(math::ArithOperator::ADD, p1, p2);";
      "p_stack->Push(res);";
      "p_stack->ReplaceWithTop(" ^ a(ins) ^ ");";
      return("1")
    ])
  );
  InstABC ( (* 35 *)
    (fun ins -> ["Sub R[%d] = R[%d] - R[%d]"; a(ins); b(ins); c(ins)]),
    (fun ins -> [
      "const auto p1 = p_stack->Get(" ^ b(ins) ^ ");";
      "const auto p2 = p_stack->Get(" ^ c(ins) ^ ");";
      "const auto res = CalcArith(math::ArithOperator::SUB, p1, p2);";
      "p_stack->Push(res);";
      "p_stack->ReplaceWithTop(" ^ a(ins) ^ ");";
      return("1")
    ])
  );
  InstABC (
    (fun ins -> ["Mul R[%d] = R[%d] * R[%d]"; a(ins); b(ins); c(ins)]),
    (fun ins -> [
      "const auto p1 = p_stack->Get(" ^ b(ins) ^ ");";
      "const auto p2 = p_stack->Get(" ^ c(ins) ^ ");";
      "const auto res = CalcArith(math::ArithOperator::MUL, p1, p2);";
      "p_stack->Push(res);";
      "p_stack->ReplaceWithTop(" ^ a(ins) ^ ");";
      return("1")
    ])
  );
  InstABC (
    (fun ins -> ["Mod R[%d] = R[%d] %% R[%d]"; a(ins); b(ins); c(ins)]),
    (fun ins -> [
      "const auto p1 = p_stack->Get(" ^ b(ins) ^ ");";
      "const auto p2 = p_stack->Get(" ^ c(ins) ^ ");";
      "const auto res = Mod(p1, p2);";
      "p_stack->Push(res);";
      "p_stack->ReplaceWithTop(" ^ a(ins) ^ ");";
      return("1")
    ])
  );
  InstABC (
    (fun ins -> ["Pow R[%d] = R[%d] ^ R[%d]"; a(ins); b(ins); c(ins)]),
    (fun ins -> [
      "const auto p1 = p_stack->Get(" ^ b(ins) ^ ");";
      "const auto p2 = p_stack->Get(" ^ c(ins) ^ ");";
      "const auto res = CalcArith(math::ArithOperator::POW, p1, p2);";
      "p_stack->Push(res);";
      "p_stack->ReplaceWithTop(" ^ a(ins) ^ ");";
      return("1")
    ])
  );
  InstABC (
    (fun ins -> ["Div R[%d] = R[%d] / R[%d]"; a(ins); b(ins); c(ins)]),
    (fun ins -> [
      "const auto p1 = p_stack->Get(" ^ b(ins) ^ ");";
      "const auto p2 = p_stack->Get(" ^ c(ins) ^ ");";
      "const auto res = CalcArith(math::ArithOperator::DIV, p1, p2);";
      "p_stack->Push(res);";
      "p_stack->ReplaceWithTop(" ^ a(ins) ^ ");";
      return("1")
    ])
  );
  InstABC ( (* 40 *)
    (fun ins -> ["I Div R[%d] = R[%d] // R[%d]"; a(ins); b(ins); c(ins)]),
    (fun ins -> [
      "const auto p1 = p_stack->Get(" ^ b(ins) ^ ");";
      "const auto p2 = p_stack->Get(" ^ c(ins) ^ ");";
      "const auto res = CalcArith(math::ArithOperator::I_DIV, p1, p2);";
      "p_stack->Push(res);";
      "p_stack->ReplaceWithTop(" ^ a(ins) ^ ");";
      return("1")
    ])
  );
  InstABC (
    (fun ins -> ["Bit And"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Bit Or"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Bit Xor"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Shift Left"]),
    (fun ins -> [return("1")])
  );
  InstABC ( (* 45 *)
    (fun ins -> ["Shift Right"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["MM Bin"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["MM Bin I"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["MM Bin K"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["UNM R[%d] = -R[%d]"; a(ins); b(ins)]),
    (fun ins -> [
      "const auto p = p_stack->Get(" ^ b(ins) ^ ");";
      "const auto res = Neg(p);";
      "p_stack->Push(res);";
      "p_stack->ReplaceWithTop(" ^ a(ins) ^ ");";
      return("1")
    ])
  );
  InstABC ( (* 50 *)
    (fun ins -> ["Bit Not"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Not R[%d] = not R[%d]"; a(ins); b(ins)]),
    (fun ins -> [
      "const auto p = p_stack->Get(" ^ b(ins) ^ ");";
      "const auto res = Not(p);";
      "p_stack->Push(res);";
      "p_stack->ReplaceWithTop(" ^ a(ins) ^ ");";
      return("1")
    ])
  );
  InstABC (
    (fun ins -> ["Len R[%d] = len(R[%d])"; a(ins); b(ins)]),
    (fun ins -> [
      "const auto p = p_stack->Get(" ^ b(ins) ^ ");";
      "const auto res = Len(p);";
      "p_stack->Push(res);";
      "p_stack->ReplaceWithTop(" ^ a(ins) ^ ");";
      return("1")
    ])
  );
  InstABC (
    (fun ins -> ["Concat R[%d] = R[%d] .. ... .. R[%d]"; a(ins); a(ins); a(ins) ^ " + " ^ b(ins) ^ " - 1"]),
    (fun ins -> [
      "auto res = p_stack->Get(" ^ a(ins) ^ ");";
      "for (int i = " ^ a(ins) ^ " + 1; i < " ^ a(ins) ^ " + " ^ b(ins) ^ "; ++i)";
      block([
        "const auto p = p_stack->Get(i);";
        "res = Concat(res, p);";
      ]);
      "p_stack->Push(res);";
      "p_stack->ReplaceWithTop(" ^ a(ins) ^ ");";
      return("1")
    ])
  );
  InstABC (
    (fun ins -> ["Close"]),
    (fun ins -> [return("1")])
  );
  InstABC ( (* 55 *)
    (fun ins -> ["TBC"]),
    (fun ins -> [return("1")])
  );
  InstsJ (
    (fun ins -> ["Jump: %d"; sj(ins)]),
    (fun ins -> [return(sj(ins) ^ " + 1")])
  );
  InstABC (
    (fun ins -> ["Equal R[%d] == R[%d] =? R[%d]"; b(ins); c(ins); a(ins)]),
    (fun ins -> [
      "const auto a = " ^ a(ins) ^ ";";
      "const auto b = p_stack->Get(" ^ b(ins) ^ ");";
      "const auto c = p_stack->Get(" ^ c(ins) ^ ");";
      return("(math::Compare(math::ComOperator::EQ, b, c) == a) ? 1 : 2")
    ])
  );
  InstABC (
    (fun ins -> ["Less Than R[%d] == R[%d] =? R[%d]"; b(ins); c(ins); a(ins)]),
    (fun ins -> [
      "const auto a = " ^ a(ins) ^ ";";
      "const auto b = p_stack->Get(" ^ b(ins) ^ ");";
      "const auto c = p_stack->Get(" ^ c(ins) ^ ");";
      return("(math::Compare(math::ComOperator::LT, b, c) == a) ? 1 : 2")
    ])
  );
  InstABC (
    (fun ins -> ["Less Equal R[%d] == R[%d] =? R[%d]"; b(ins); c(ins); a(ins)]),
    (fun ins -> [
      "const auto a = " ^ a(ins) ^ ";";
      "const auto b = p_stack->Get(" ^ b(ins) ^ ");";
      "const auto c = p_stack->Get(" ^ c(ins) ^ ");";
      return("(math::Compare(math::ComOperator::LE, b, c) == a) ? 1 : 2")
    ])
  );
  InstABC ( (* 60 *)
    (fun ins -> ["Equal K"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Equal I"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Less Than I"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Less Equal I"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Greater Than I"]),
    (fun ins -> [return("1")])
  );
  InstABC ( (* 65 *)
    (fun ins -> ["Greater Equal I"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Test: if (R[%d] != %d) then pc++"; a(ins); c(ins)]),
    (fun ins -> [
      "const auto a = p_stack->Get(" ^ a(ins) ^ ");";
      "const auto c = " ^ c(ins) ^ ";";
      return("(ToBoolean(a) != ToBoolean(c)) ? 2 : 1")
    ])
  );
  InstABC (
    (fun ins -> ["Test Set: if (R[%d] != %d) then R[%d] = R[%d] else pc++"; b(ins); c(ins); a(ins); b(ins)]),
    (fun ins -> [
      "const auto b = p_stack->Get(" ^ b(ins) ^ ");";
      "const auto c = " ^ c(ins) ^ ";";
      "if (ToBoolean(b) != ToBoolean(c))";
      block([
        "p_stack->Set(" ^ a(ins) ^ ", b);";
        return("1");
      ]);
      "else";
      block([return("2")])
    ])
  );
  InstABC (
    (fun ins -> ["Call"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Tail Call"]),
    (fun ins -> [return("1")])
  );
  InstABC ( (* 70 *)
    (fun ins -> ["Return: %d, %d"; (a ins); (b ins)]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Return 0"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Return 1"]),
    (fun ins -> [return("1")])
  );
  InstABx (
    (fun ins -> ["For Loop: R[%d] += R[%d]; if R[%d] <= R[%d] then pc -= %d";
      a(ins); a(ins) ^ " + 2"; a(ins); a(ins) ^ " + 1"; bx(ins)]),
    (fun ins -> [
      "UpdateForCount(p_stack, " ^ a(ins) ^ ");";
      return("(ShouldSkipForLoop(p_stack, " ^ a(ins) ^ ")) ? 1 : 1 - " ^ bx(ins))
    ])
  );
  InstABx (
    (fun ins -> ["For Prepare: if R[%d] > R[%d] pc += %d"; a(ins); a(ins) ^ " + 1"; bx(ins)]),
    (fun ins -> [return("(ShouldSkipForLoop(p_stack, " ^ a(ins) ^ ")) ? 1 : 1 + " ^ bx(ins))])
  );
  InstABx ( (* 75 *)
    (fun ins -> ["T For Prepare"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["T For Call"]),
    (fun ins -> [return("1")])
  );
  InstABx (
    (fun ins -> ["T For Loop"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Set List"]),
    (fun ins -> [return("1")])
  );
  InstABx (
    (fun ins -> ["Closure"]),
    (fun ins -> [return("1")])
  );
  InstABC ( (* 80 *)
    (fun ins -> ["Variable Arguments"]),
    (fun ins -> [return("1")])
  );
  InstABC (
    (fun ins -> ["Variable Arguments Prepare"]),
    (fun ins -> [return("1")])
  );
  InstAx (
    (fun ins -> ["Extra Arguments"]),
    (fun ins -> [return("1")])
  );
];;

let pretty_print_cpp = (pretty_print
  (List.map (fun ins -> match ins with
    | InstABC(pp, _) as ins -> pp ins
    | InstABx (pp, _) as ins -> pp ins
    | InstAsBx(pp, _) as ins -> pp ins
    | InstAx(pp, _) as ins -> pp ins
    | InstsJ(pp, _) as ins -> pp ins
  ) instructions));;

let impl_cpp = (execute
(List.map (fun ins -> match ins with
  | InstABC(_, stmts) as ins -> stmts ins
  | InstABx (_, stmts) as ins -> stmts ins
  | InstAsBx(_, stmts) as ins -> stmts ins
  | InstAx(_, stmts) as ins -> stmts ins
  | InstsJ(_, stmts) as ins -> stmts ins
) instructions));;

let filename = "test/cpp/instructions_pretty_printer.cc" in
  let oc = open_out filename in
    Printf.fprintf oc "%s\n" pretty_print_cpp;
    close_out oc;;

let filename = "main/luna/instructions/instructions_impl.cc" in
  let oc = open_out filename in
    Printf.fprintf oc "%s\n" impl_cpp;
    close_out oc;;
